internal/ts_library/rule.bzl

```diff
@@ -1,286 +1,102 @@
 load("//internal/js_library:rule.bzl", "JsLibraryInfo")
 load("//internal/npm_packages:rule.bzl", "NpmPackagesInfo")
+load("//internal/common:context.bzl", "JS_LIBRARY_ATTRIBUTES", "js_context")
 
-TsLibraryInfo = provider(fields=[
-  # Directory containing the TypeScript files (and potentially other assets).
-  "original_typescript_dir",
-  # Source files provided as input.
-  "typescript_source_files",
-  # Directory containing the generated TypeScript definitions and compiled JavaScript.
-  "compiled_typescript_dir",
+TsLibraryInfo = provider(fields = [
+    # Directory containing the TypeScript files (and potentially other assets).
+    "original_typescript_dir",
+    # Source files provided as input.
+    "typescript_source_files",
+    # Directory containing the generated TypeScript definitions and compiled JavaScript.
+    "compiled_typescript_dir",
 ])
 
 def _ts_library_impl(ctx):
-  # Ensure that we depend on at most one npm_packages, since we don't want to
-  # have conflicting package versions coming from separate node_modules
-  # directories.
-  direct_npm_packages = [
-    dep
-    for dep in ctx.attr.deps
-    if NpmPackagesInfo in dep
-  ]
-  if len(direct_npm_packages) > 1:
-    fail("Found more than one set of NPM packages in target definition: " + ",".join([
-      dep.label
-      for dep in direct_npm_packages
-    ]))
-  extended_npm_packages = depset(
-    direct = direct_npm_packages,
-    transitive = [
-      dep[JsLibraryInfo].npm_packages
-      for dep in ctx.attr.deps
-      if JsLibraryInfo in dep
-    ],
-  )
-  npm_packages_list = extended_npm_packages.to_list()
-  if len(npm_packages_list) > 1:
-    fail("Found more than one set of NPM packages through dependencies: " + ",".join([
-      dep.label
-      for dep in npm_packages_list
-    ]))
-  # If we depend on an npm_packages target, we'll use its node_modules
-  # directory to find modules. Otherwise, we'll use an empty node_modules
-  # directory.
-  npm_packages = (
-    npm_packages_list[0] if len(npm_packages_list) == 1
-    else ctx.attr._empty_npm_packages
-  )
-  # Gather all internal deps (other ts_library rules).
-  internal_deps = depset(
-    direct = [
-      dep
-      for dep in ctx.attr.deps
-      if JsLibraryInfo in dep
-    ],
-    transitive = [
-      dep[JsLibraryInfo].internal_deps
-      for dep in ctx.attr.deps
-      if JsLibraryInfo in dep
-    ],
-  )
+    js = js_context(ctx)
+    providers = []
+    # for i, d in enumerate(ctx.attr.deps):
```

Comment:
probably good to remove now?

Reply:
I assume this is pending a commit that hasn't been pushed yet?

Reply:
It's pushed.